### PR TITLE
【Hackathon No.30】

### PR DESCRIPTION
**Pull request by yangguohao**
Create api_design_for_tripletmargindistanceloss.md

### CODE REVIEW

**Comment by dingjiaweiww** [comment]
你的 PR 提交成功，感谢你对于开源项目的贡献，请检查 PR 提交格式和内容是否完备，具体请参考[示例](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20200301_api_design_for_quantile.md)和[模版](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/api_design_template.md)。

**Comment by dingjiaweiww** [comment]
PR 格式检查通过，你的PR 将接受Paddle 专家以及开源社区的review，请及时关注PR 动态

File: rfcs/APIs/api_design_for_tripletmargindistanceloss.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+为了提升飞桨API丰富度，支持科学计算领域API，Paddle需要扩充APIpaddle.nn.TripleMarginDistanceLoss以及paddle.nn.functional.triplet_margin_with_distance_loss
+## 2、功能目标
+paddle.nn.TripletMarginDistanceLoss 是三元损失函数，其针对 anchor 和正负对计算 P 范数距离下的三元损失，从而获得损失值。
+## 3、意义
+为 paddle 框架中新增计算损失函数的方法
+
+# 二、飞桨现状
+对飞桨框架目前支持此功能的现状调研，如果不支持此功能，如是否可以有替代实现的API，是否有其他可绕过的方式，或者用其他API组合实现的方式；
+目前paddle缺少相关功能实现。
+需要独立设计实现相关的函数
```
Comment (shiyutang) [inline rfcs/APIs/api_design_for_tripletmargindistanceloss.md:24]: 说明语言可以去除，并且可以对相关替代方式进行说明

File: rfcs/APIs/api_design_for_tripletmargindistanceloss.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+> Creates a criterion that measures the triplet loss given an input tensors $x 1, x 2, x 3$ and a margin with a value greater than 0 . This is used for measuring a relative similarity between samples. A triplet is composed by a, $p$ and $n$ (i.e., anchor, positive examples and negative examples respectively). The shapes of all input tensors should be $(N, D)$.
+>
+> The distance swap is described in detail in the paper Learning shallow convolutional feature descriptors with triplet losses by V. Balntas, E. Riba et al.
+> The loss function for each sample in the mini-batch is:
+>
+> $$
+> L(a, p, n)=\max \left\{d\left(a_{i}, p_{i}\right)-d\left(a_{i}, n_{i}\right)+\operatorname{margin}, 0\right\}
+> $$
+>
```
Comment (shiyutang) [inline rfcs/APIs/api_design_for_tripletmargindistanceloss.md:39]: 这部分需要有关键代码和方案分析：https://github.com/pytorch/pytorch/blob/cfb6c942fed64dbb81ccc4f14b2a6650123af2e1/torch/nn/functional.py#L4564
另外tf的方案也可以寻找一下并同样说明

File: rfcs/APIs/api_design_for_tripletmargindistanceloss.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+
+# 四、对比分析
+paddle 和 pytorch整体框架相似，故直接采用其方法进行设计。
```
Comment (shiyutang) [inline rfcs/APIs/api_design_for_tripletmargindistanceloss.md:43]: 对比分析需要对比方案和选择原因

File: rfcs/APIs/api_design_for_tripletmargindistanceloss.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+- `paddle.nn.TripletMarginDistanceLoss(margin=1.0, distance_function=None, swap=False, reduction='mean', name=None) -> Tensor`
+- `padde.nn.functional.triplet_margin_with_distance_loss(input, positive, negative, distance_function=None, margin=1.0, swap=False, reduction='mean', name=None) -> Tensor`
+## 底层OP设计
+## API实现方案
+distance functions可以采用paddle.nn.PairWiseDistance来进行实现
+1. 检查参数
+
+   1. 检查 reduction 有效性（同其余 functional loss 中的实现）
+   2. 检查输入的 dtype（含 `input`、`positive`、`negative`）（同其余 functional loss 中的实现）
+
```
Comment (shiyutang) [inline rfcs/APIs/api_design_for_tripletmargindistanceloss.md:60]: 检查输入的维度一致性

**Review by shiyutang — CHANGES_REQUESTED** [review]
存在较多需要修正之处，示例可以参考：https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20200301_api_design_for_quantile.md

**Review by TCChenlong — CHANGES_REQUESTED** [review]
@yangguohao 已review哈，请按上述描述优化你的RFC，感谢~

File: rfcs/APIs/api_design_for_tripletmargindistanceloss.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+- `paddle.nn.TripletMarginDistanceLoss(margin=1.0, distance_function=None, swap=False, reduction='mean', name=None) -> Tensor`
+- `padde.nn.functional.triplet_margin_with_distance_loss(input, positive, negative, distance_function=None, margin=1.0, swap=False, reduction='mean', name=None) -> Tensor`
+## 底层OP设计
+## API实现方案
+distance functions可以采用paddle.nn.PairWiseDistance来进行实现
+1. 检查参数
+    
+   1. 检查 reduction 有效性（同其余 functional loss 中的实现）
+   2. 检查输入的 dtype（含 `input`、`positive`、`negative`）（同其余 functional loss 中的实现）
```
Comment (shiyutang) [inline rfcs/APIs/api_design_for_tripletmargindistanceloss.md:166]: 输入损失的不一定为（N，dim），需要检查维度之后进行维度转换

File: rfcs/APIs/api_design_for_tripletmargindistanceloss.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+   3. 检查输入的`input`、`positive`、`negative`维度是否相同
+
+2. 计算
+
+   1. 用户可传入distance_function参数，如果未指定则使用 `paddle.nn.PairWiseDistance` 分别计算得到正锚点与样本和负锚点与样本的距离。
```
Comment (shiyutang) [inline rfcs/APIs/api_design_for_tripletmargindistanceloss.md:171]: 可以替换为paddle.linalg.pnorm

File: rfcs/APIs/api_design_for_tripletmargindistanceloss.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+3. 根据 `reduction`，输出 loss（同其余 functional loss 中的实现）
+# 六、测试和验收的考量
+
+测试考虑的case如下:
+- 1.动态图，静态图，要与np计算下的结果输出需要一致。
+- 2.CPU、GPU下计算一致。
+- 3.各reduction下计算一致
+- 4.各参数输入有效。
+- 5.反向梯度的正确性。
```
Comment (shiyutang) [inline rfcs/APIs/api_design_for_tripletmargindistanceloss.md:183]: 目前已经调整基于python 进行组合的API不需要进行反向梯度的验证

**Review by shiyutang — CHANGES_REQUESTED** [review]
部分实现细节需要修改

**Review by shiyutang — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]
